### PR TITLE
Michaelw/hash images

### PR DESCRIPTION
**Pull request by mstbbs**
## What is it?
fingerprint images with their md5 hash. 
also move cache busting for css and js into gulp.

**Images have moved from the `static/` folder to the `src/` folder.**

The largest change is that the site won't load images using native hugo commands `hugo server --renderToDisk`, unless you have also run `gulp build` or `gulp watch`. This is because all images have now been moved to /src folder.

## Preview Site

https://d2cx6t4ssmwdrv.cloudfront.net/michaelw/hash-images/guides/hostmap/

Look at all those nicely hashed image names!

### CODE REVIEW

File: content/integrations/azure.ja.md [hunk 1]
```diff
@@ -71,7 +71,8 @@ Datadogの監視が正しく動作するためには、"Basic metrics", "network
 Datadogは、Azure Extension Managerに対応しています。これにより、AzureのポータルからDatadog Agentをインストールすることが出来ます。Azure上で起動しているVMがWindowsかLinuxに関わらず、ワンクリックでDatadog Agentをインストールすることが出来ます。ワンクリックでのインストールを実行するには、[
 Azureのポータル](http://portal.azure.com)から、運用中のVMを選択するか、新しいVMを起動しDatadog Agentの追加をクリックします。
 
-![Azure performance monitoring with One-Click Deployment through Datadog](/images/azure-image-4.gif)
+{{< img src="azure-image-4.gif" alt="Azure performance monitoring with One-Click Deployment through Datadog" >}}
+
```
Comment (mstbbs) [inline content/integrations/azure.ja.md]: shortcode would now be required.

File: layouts/404.html [hunk 2]
```diff
@@ -8,8 +8,7 @@
         <div class="col-xs-12 col-sm-9 col-md-9 main">
             <h1>Uh oh, Something Went Wrong</h1>
 
-            <img src="{{ .Site.Params.img_url }}images/sad_bits.png" />
-
+            <img src="{{ partial "img.html" (dict "root" . "src" "sad_bits.png") }}" />
```
Comment (mstbbs) [inline layouts/404.html]: new partial will also be required for images

File: layouts/partials/footer.html [hunk 3]
```diff
@@ -28,7 +28,7 @@
           </script>
         {{ end }}
 
-        <script src="{{ .Site.BaseURL }}js/main-dd.js"></script>
+        <script src="{{ .Site.BaseURL }}js/{{ (index .Site.Data.manifests.js "main-dd.js") }}"></script>
```
Comment (mstbbs) [inline layouts/partials/footer.html]: static assets are loaded from a manifest which will map friendly name to hashed name

File: layouts/partials/img.html [hunk 4]
```diff
@@ -0,0 +1 @@
+{{ .root.Site.Params.img_url }}images/{{ (index .root.Site.Data.manifests.images .src ) }}
```
Comment (mstbbs) [inline layouts/partials/img.html]: load image from manifest to get the correct file name

File: layouts/partials/sidebar.html [hunk 5]
```diff
@@ -13,7 +13,9 @@
     <li class="nav-header">Documentation Search</li>
     <li>
       <form action="{{ $.Site.BaseURL }}search/">
-      <input type="text" name="q" id="tipue_search_input" autocomplete="off" required>
+      <input type="text" name="q" id="tipue_search_input" autocomplete="off"
+             style="background-image: url({{ partial "img.html" (dict "root" . "src" "search.png") }});"
+             required>
```
Comment (mstbbs) [inline layouts/partials/sidebar.html]: background images cannot be loaded in css

File: local/bin/sh/helpers.sh [hunk 6]
```diff
@@ -7,15 +7,12 @@ version_static_assets() {
...
+        npm install  # make sure everything is uptodate
         if [[ "${BUCKET}" == *"preview"* ]]; then
             gulp build || fail_step "${FUNCNAME}"
         else
             gulp build --production || fail_step "${FUNCNAME}"
         fi
-        # ============ version created assets ============ #
-        echo "--------"
-        echo "Fingerprinting static assets... "
-        version_static.py || fail_step "${FUNCNAME}"
```
Comment (mstbbs) [inline local/bin/sh/helpers.sh]: We no longer need this

File: local/bin/sh/run-site.sh [hunk 7]
```diff
@@ -14,6 +14,7 @@ if [ ${RUN_SERVER} == true ]; then
 		echo "checking that node modules are installed and up-to-date"
 		npm install
         echo "starting gulp watch"
+        gulp build
```
Comment (mstbbs) [inline local/bin/sh/run-site.sh]: we need to make sure we build every startup, which will create any missing manifests

**Comment by l0k0ms** [comment]
One side effect though, we can not build the doc locally with "make docker-start" now.

**Pull request merged**